data/transactions/logic/pairing_test.go

```diff
@@ -16,14 +16,543 @@
+	bn254fp "github.com/consensys/gnark-crypto/ecc/bn254/fp"
+)
+
 const pairingNonsense = `
  pushbytes 0x012345
  dup
- bn256_add
+ ec_add BN254g1
  dup
- bn256_scalar_mul
+ ec_scalar_mul BLS12_381g2
  dup
- bn256_pairing
+ ec_pairing_check BN254g1
+ ec_multi_exp BLS12_381g2
+ ec_subgroup_check BLS12_381g1
+ ec_map_to BN254g2
 `
 
-const pairingCompiled = "80030123454999499a499b"
+const pairingCompiled = "800301234549e00049e10349e200e303e402e501"
+
+func bn254G1sToBytes(g1s []bn254.G1Affine) []byte {
+	var out []byte
+	for i := range g1s {
+		out = append(out, bn254G1ToBytes(&g1s[i])...)
+	}
+	return out
+}
+
+func bn254G2sToBytes(g2s []bn254.G2Affine) []byte {
+	var out []byte
+	for i := range g2s {
+		out = append(out, bn254G2ToBytes(&g2s[i])...)
+	}
+	return out
+}
+
+func bls12381G1sToBytes(g1s []bls12381.G1Affine) []byte {
+	var out []byte
+	for i := range g1s {
+		out = append(out, bls12381G1ToBytes(&g1s[i])...)
+	}
+	return out
+}
+
+func bls12381G2sToBytes(g2s []bls12381.G2Affine) []byte {
+	var out []byte
+	for i := range g2s {
+		out = append(out, bls12381G2ToBytes(&g2s[i])...)
+	}
+	return out
+}
+
+func TestEcAdd(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// Parameterize this to check all curves
+
+	g1point := bn254RandomG1()
+	g1bytes1 := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1ToBytes(&g1point)))
+	g1point = bn254RandomG1()
+	g1bytes2 := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1ToBytes(&g1point)))
+	testAccepts(t, g1bytes1+g1bytes2+"ec_add BN254g1; len", pairingVersion)
+	testAccepts(t, g1bytes1+"int 64; bzero; ec_add BN254g1;"+g1bytes1+"==", pairingVersion)
+}
+
+func TestEcScalarMul(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	// Parameterize this to check all curves
+
+	g1point := bn254RandomG1()
+	g1bytes1 := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1ToBytes(&g1point)))
+	testAccepts(t, g1bytes1+"int 0; itob; ec_scalar_mul BN254g1; int 64; bzero; ==", pairingVersion)
+	testPanics(t, g1bytes1+"int 33; bzero; ec_scalar_mul BN254g1; int 64; bzero; ==", pairingVersion, "ec_scalar_mul scalar len is 33")
+	testAccepts(t, g1bytes1+"int 1; itob; ec_scalar_mul BN254g1;"+g1bytes1+"==", pairingVersion)
+	testAccepts(t, g1bytes1+`
+dup
+int 32; bzero; int 7; itob; b|
+ec_scalar_mul BN254g1
+swap
+int 7; itob
+ec_scalar_mul BN254g1
+==
+`, pairingVersion)
+}
+
+func TestPairCheck(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	t.Parallel()
+
+	t.Run("bn254", func(t *testing.T) {
+		t.Parallel()
+		var g1GenNeg bn254.G1Affine
+		g1GenNeg.Neg(&bnG1Gen)
+		g1points := []bn254.G1Affine{g1GenNeg, bnG1Gen}
+		g2points := []bn254.G2Affine{bnG2Gen, bnG2Gen}
+		// -1 g1 g2 + g1 g2 = 0
+		g1bytes := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G1sToBytes(g1points)))
+		g2bytes := fmt.Sprintf("byte 0x%s\n", hex.EncodeToString(bn254G2sToBytes(g2points)))
+
+		testAccepts(t, g1bytes+g2bytes+`ec_pairing_check BN254g1`, pairingVersion)
```

Comment:
I did not realize that you can specify a group G1/G2 for pairing and that switches the two arguments.
I think this is a bit unusual (though not a security issue).
Would is be possible to only allow `BN254` not `BN254g1/g2` and then requires g1 is always first?

Reply:
That's what I had, but creates a bunch of documentation and wording questions because "BN254" is not a curve.  We were all annoyed at need to say that some of the opcodes worked on the "...G1" curves/groups, but not the unsuffixed names, etc

So do we create to different sets of constants? The g1/g2 table, which might be called "curves" and then the table that does not have the suffixes and call the "curve families"?  

Seems like this sidesteps all of that and only talks about the actual curves, the G1/G2s.

Reply:
I cannot imagine what questions can possibly arise by insisting on g1-then-g2 ordering.

In general I like the proposal of calling it "BN12-384 family" when referring to the entire system, and talk about G1, G2 when needing to specify a particular group. If there is ever a need to talk about the curves themselves (which I hope there isn't), then call them "curve 1" and "curve 2"